src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java

```diff
@@ -86,51 +98,79 @@ public DeprovisionWorkflowTransportAction(
     }
 
     @Override
     protected void doExecute(Task task, WorkflowRequest request, ActionListener<WorkflowResponse> listener) {
         String workflowId = request.getWorkflowId();
-        String allowDelete = request.getParams().get(ALLOW_DELETE);
-        GetWorkflowStateRequest getStateRequest = new GetWorkflowStateRequest(workflowId, true);
+
+        User user = getUserContext(client);
 
         // Stash thread context to interact with system index
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-            logger.info("Querying state for workflow: {}", workflowId);
-            client.execute(GetWorkflowStateAction.INSTANCE, getStateRequest, ActionListener.wrap(response -> {
-                context.restore();
+            resolveUserAndExecute(
+                user,
+                workflowId,
+                filterByEnabled,
+                listener,
+                () -> executeDeprovisionRequest(request, listener, context),
+                client,
+                clusterService,
+                xContentRegistry
+            );
 
-                Set<String> deleteAllowedResources = Strings.tokenizeByCommaToSet(allowDelete);
-                // Retrieve resources from workflow state and deprovision
-                threadPool.executor(DEPROVISION_WORKFLOW_THREAD_POOL)
-                    .execute(
-                        () -> executeDeprovisionSequence(
-                            workflowId,
-                            response.getWorkflowState().resourcesCreated(),
-                            deleteAllowedResources,
-                            listener
-                        )
-                    );
-            }, exception -> {
-                String errorMessage = "Failed to get workflow state for workflow " + workflowId;
-                logger.error(errorMessage, exception);
-                listener.onFailure(new FlowFrameworkException(errorMessage, ExceptionsHelper.status(exception)));
-            }));
         } catch (Exception e) {
             String errorMessage = "Failed to retrieve template from global context.";
             logger.error(errorMessage, e);
             listener.onFailure(new FlowFrameworkException(errorMessage, ExceptionsHelper.status(e)));
         }
     }
 
+    private void executeDeprovisionRequest(
+        WorkflowRequest request,
+        ActionListener<WorkflowResponse> listener,
+        ThreadContext.StoredContext context
+    ) {
+        String workflowId = request.getWorkflowId();
+        String allowDelete = request.getParams().get(ALLOW_DELETE);
+        GetWorkflowStateRequest getStateRequest = new GetWorkflowStateRequest(workflowId, true);
+        logger.info("Querying state for workflow: {}", workflowId);
+        client.execute(GetWorkflowStateAction.INSTANCE, getStateRequest, ActionListener.wrap(response -> {
+            context.restore();
```

Comment:
Not sure the context is restored in the exception path here?  (We're relying on the try-with-resource wrapping this call.)   Consider using `runBefore()`.

Reply:
> It should be. Since it was passed from a try/catch block above. 

Right, the try-with-resources block will close it it but it seems inconsistent with explicitly restoring it here on this path.  Mostly just a consistency/readability issue.